File: C++/GTKmm/src/widgets/todo_list.cpp
#include "todo_list.h"
#include "../theme.h"
#include "../cairo_util.h"
#include <pangomm.h>
#include <cmath>

// ── TodoItemWidget ────────────────────────────────────────────────────────────

TodoItemWidget::TodoItemWidget(const Todo& todo) : todo_(todo) {
    set_size_request(-1, Theme::ITEM_H);
    add_events(Gdk::BUTTON_PRESS_MASK);
}

bool TodoItemWidget::on_draw(const Cairo::RefPtr<Cairo::Context>& cr) {
    int w = get_allocated_width();
    int h = get_allocated_height();

    set_color(cr, Theme::SURFACE);
    rounded_rect(cr, 0, 0, w, h, 4);
    cr->fill();

    double cx  = Theme::CB_PAD_L + Theme::CB_SIZE / 2.0;
    double cy  = h / 2.0;
    double r   = Theme::CB_SIZE / 2.0;

    if (todo_.is_completed) {
        set_color(cr, Theme::ACCENT);
        cr->arc(cx, cy, r, 0, 2 * M_PI);
        cr->fill();
        set_color(cr, Theme::WHITE);
        cr->set_line_width(1.5);
        cr->set_line_cap(Cairo::LINE_CAP_ROUND);
        cr->set_line_join(Cairo::LINE_JOIN_ROUND);
        cr->move_to(cx - 4, cy);
        cr->line_to(cx - 1, cy + 3);
        cr->line_to(cx + 4, cy - 3);
        cr->stroke();
    } else {
        set_color(cr, Theme::CB_BDR);
        cr->set_line_width(2.0);
        cr->arc(cx, cy, r - 1, 0, 2 * M_PI);
        cr->stroke();
    }

    int tx = Theme::CB_PAD_L + Theme::CB_SIZE + Theme::CB_MR;
    set_color(cr, todo_.is_completed ? Theme::MUTED : Theme::TEXT_C);

    auto layout = create_pango_layout(todo_.task);
    layout->set_font_description(Pango::FontDescription("Sans 10"));

    if (todo_.is_completed) {
        auto attr_list = Pango::AttrList();
        auto strike = Pango::Attribute::create_attr_strikethrough(true);
        attr_list.insert(strike);
        layout->set_attributes(attr_list);
    }

    int text_w, text_h;
    layout->get_pixel_size(text_w, text_h);
    cr->move_to(tx, (h - text_h) / 2.0);
    layout->show_in_cairo_context(cr);

    return true;
}

bool TodoItemWidget::on_button_press_event(GdkEventButton* ev) {
    double cx_start = Theme::CB_PAD_L;
    double cx_end   = Theme::CB_PAD_L + Theme::CB_SIZE;
    double cy_start = (Theme::ITEM_H - Theme::CB_SIZE) / 2.0;
    double cy_end   = cy_start + Theme::CB_SIZE;

    if (ev->x >= cx_start && ev->x <= cx_end &&
        ev->y >= cy_start && ev->y <= cy_end) {
        signal_toggle.emit(todo_.id);
    }
    return true;
}

// ── TodoList ──────────────────────────────────────────────────────────────────

TodoList::TodoList() {
    set_policy(Gtk::POLICY_NEVER, Gtk::POLICY_AUTOMATIC);
    set_hexpand(true);
    set_vexpand(true);
    vbox_.set_spacing(0);
    add(vbox_);
}

void TodoList::refresh(const std::vector<Todo>& todos) {
    for (auto* w : vbox_.get_children())
        vbox_.remove(*w);
    items_.clear();

    for (const auto& t : todos) {
        auto* item = Gtk::make_managed<TodoItemWidget>(t);
        item->signal_toggle.connect([this](int id){ signal_toggle.emit(id); });
        item->set_margin_bottom(Theme::ITEM_MB);
        vbox_.pack_start(*item, false, false, 0);
        items_.push_back(item);
    }
    vbox_.show_all();
}

File: Python/PyGTK/widgets/todo_list.py
import gi
gi.require_version("Gtk", "3.0")
gi.require_version("Pango", "1.0")
gi.require_version("PangoCairo", "1.0")
from gi.repository import Gtk, Gdk, Pango, PangoCairo
import cairo
import math
from typing import List, Callable
from data import Todo
from theme import Theme, set_color, rounded_rect


class TodoItemWidget(Gtk.DrawingArea):
    def __init__(self, todo: Todo, on_toggle: Callable[[int], None]):
        super().__init__()
        self._todo = todo
        self._on_toggle = on_toggle
        self.set_size_request(-1, Theme.ITEM_H)
        self.add_events(Gdk.EventMask.BUTTON_PRESS_MASK)
        self.connect("draw", self._on_draw)
        self.connect("button-press-event", self._on_click)

    def _on_draw(self, _widget, cr):
        w = self.get_allocated_width()
        h = self.get_allocated_height()

        # Background
        set_color(cr, Theme.SURFACE_C)
        rounded_rect(cr, 0, 0, w, h, 4)
        cr.fill()

        # Checkbox
        cx = Theme.CB_PAD_L + Theme.CB_SIZE / 2.0
        cy = h / 2.0
        r = Theme.CB_SIZE / 2.0

        if self._todo.is_completed:
            set_color(cr, Theme.CB_BDR_C)
            cr.set_line_width(2.0)
            cr.arc(cx, cy, r - 1, 0, 2 * math.pi)
            cr.stroke()
            set_color(cr, Theme.ACCENT_C)
            cr.set_line_width(1.5)
            cr.set_line_cap(cairo.LINE_CAP_ROUND)
            cr.set_line_join(cairo.LINE_JOIN_ROUND)
            cr.new_path()
            cr.move_to(cx - 4, cy)
            cr.line_to(cx - 1, cy + 3)
            cr.line_to(cx + 4, cy - 3)
            cr.stroke()
        else:
            set_color(cr, Theme.CB_BDR_C)
            cr.set_line_width(2)
            cr.arc(cx, cy, r - 1, 0, 2 * math.pi)
            cr.stroke()

        # Text
        tx = Theme.CB_PAD_L + Theme.CB_SIZE + Theme.CB_MR
        layout = PangoCairo.create_layout(cr)
        layout.set_text(self._todo.task, -1)
        desc = Pango.FontDescription.from_string("Sans 10")
        layout.set_font_description(desc)

        if self._todo.is_completed:
            attrs = Pango.AttrList()
            strike = Pango.attr_strikethrough_new(True)
            attrs.insert(strike)
            layout.set_attributes(attrs)

        color = Theme.MUTED_C if self._todo.is_completed else Theme.TEXT_C
        set_color(cr, color)
        _ink, logical = layout.get_pixel_extents()
        ty = (h - logical.height) / 2.0
        cr.move_to(tx, ty)
        PangoCairo.show_layout(cr, layout)

    def _on_click(self, _widget, event):
        cx = Theme.CB_PAD_L
        cy = (Theme.ITEM_H - Theme.CB_SIZE) / 2.0
        if cx <= event.x <= cx + Theme.CB_SIZE and cy <= event.y <= cy + Theme.CB_SIZE:
            self._on_toggle(self._todo.id)


class TodoList(Gtk.ScrolledWindow):
    def __init__(self, on_toggle: Callable[[int], None]):
        super().__init__()
        self._on_toggle = on_toggle
        self.set_policy(Gtk.PolicyType.NEVER, Gtk.PolicyType.AUTOMATIC)
        self.set_vexpand(True)

        self._vbox = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=Theme.ITEM_MB)
        self._vbox.set_valign(Gtk.Align.START)
        self.add(self._vbox)

    def refresh(self, todos: List[Todo]):
        for child in self._vbox.get_children():
            self._vbox.remove(child)
        for todo in todos:
            item = TodoItemWidget(todo, self._on_toggle)
            self._vbox.pack_start(item, False, False, 0)
        self._vbox.show_all()